keps/sig-testing/1746-stage-e2e-test-framework/README.md

```diff
@@ -0,0 +1,610 @@
+
+### Goals
+
+<!--
+List the specific goals of the KEP.  What is it trying to achieve?  How will we
+know that this has succeeded?
+-->
+
+* Decouple the e2e test framework and all of its subpackages from any
+  Kubernetes internal packages (i.e., pkg, other e2e packages).
+* Impose import restrictions to prevent the use of core Kubernetes packages.
+* Allow the e2e test framework to be easily used and without requiring the
+  entirety of Kubernetes to be imported.
+* Refactor the e2e test framework so that its sub-packages are well contained
+  (i.e., `framework/pod/*` contains all necessary utilities to manage pods
+  during e2e tests).
+  * Update Kubernetes e2e tests to use the refactored version of any utility
+    function we change.
+
+### Non-Goals
+
+* Change the way e2e tests/jobs are setup or executed.
+* This enhancement is not directed at making the e2e test framework
+  provider-agnostic. Cloud provider functionality will be kept as is unless it
+  is used in inadequate e2e test framework sub-packages.
+* The e2e test framework will be refactored as needed strictly to remove its
+  dependence from other internal Kubernetes packages. It is not the goal of
+  this enhancement to propose an entire redesign of the e2e test framework.
+
+<!--
+What is out of scope for this KEP?  Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+In order to cleanup and refactor the e2e test framework we will have to
+decouple the e2e test framework from internal Kubernetes packages and to
+organize test code in e2e tests and in the e2e test framework.
+
+### User Stories
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system.  The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+I, as a core Kubernetes contributor, want a e2e test framework that is well
+organized and containsa core of functions that will help me write clean and
+maintainable e2e tests.
+
+#### Story 2
+
+I, as a maintainer of a project running on top of Kubernetes (e.g. external cloud providers, CSI drivers), want to use the
+same infrastructure and tools used by core Kubernetes without having to import
+core Kubernetes.
+
+### Managing Internal Kubernetes Packages
+
+To decouple the e2e test framework from internal packages we essentially have
+to prevent any from `k8s.io/kubernetes/pkg` and from any other package in
+`k8s.io/kubernetes/test` that aren't under
+`k8s.io/kubernetes/test/e2e/framework`.
+
+These core dependencies are introduced in the following ways
```

Comment:
these are substantially introduced via the "provider" concept throughout the existing tests.

Reply:
It depends on the context, the vast majority of refactoring testing commons has done for these areas has been based on the utility functions that rely on internal methods and variables of Kubernetes.